frontend/src/Theme/ThemeProvider.tsx

```diff
@@ -35,77 +35,87 @@ type ThemeProviderProps = {
   children: ReactNode
 }
 
-export function ThemeProvider(props: ThemeProviderProps) {
-  const [theme, setThemeActual] = useState<string>()
-  const [currentStyles, setCurrentStylesActual] = useState<{ styles: ThemeStyles; withTransition: boolean }>()
-
-  useEffect(() => {
-    // Restore theme and styles from localStorage
-    let { theme, styles } = restoreTheme()
-
-    if (!theme) {
-      theme = props.initialTheme || 'light'
-    }
-
-    if (theme) {
-      setThemeActual(theme)
-    }
-
-    // Don't override predefined themes
-    if (props.themeCollection[theme]) {
-      styles = props.themeCollection[theme]
+const globalStylesheet = (() => {
+  const head = document.head || document.getElementsByTagName('head')[0]
+  const style = document.createElement('style')
+  head.appendChild(style)
+  style.appendChild(document.createTextNode(''))
+  return style
+})()
+
+export const ThemeProvider = observer((props: ThemeProviderProps) => {
+  const appState = useAppState()
+  const [currentStyles, setCurrentStylesActual] = useMemo(() => {
+    // Use a ref to track if we're in a transition
+    const state = { styles: undefined as ThemeStyles | undefined, withTransition: false }
+
+    const setCurrentStyles = (styles: ThemeStyles, withTransition = true) => {
+      state.styles = styles
+      state.withTransition = withTransition
+      applyTheme(globalStylesheet, styles, withTransition)
     }
 
-    setCurrentStylesActual({ styles, withTransition: false })
-  }, [props.themeCollection, props.initialTheme])
-
-  const stylesheet = useMemo(() => {
-    const head = document.head || document.getElementsByTagName('head')[0]
-    const style = document.createElement('style')
-    head.appendChild(style)
-    style.appendChild(document.createTextNode(''))
-    return style
+    return [state, setCurrentStyles] as const
   }, [])
 
-  const { setTheme, setCurrentStyles } = useMemo(() => {
-    const setTheme = (newTheme: string) => {
-      setThemeActual(newTheme)
-      if (props.themeCollection[newTheme]) {
-        setCurrentStyles(props.themeCollection[newTheme])
+  // Initialize theme from AppState
+  useEffect(() => {
+    let theme = props.initialTheme || appState.theme
+    let styles = props.themeCollection[theme]
+
+    if (!styles) {
+      // Try to use preferred color scheme as fallback
+      const preferredTheme = getPreferredColorScheme()
```

Comment:
maybe move this to AppState constructor to have theme initialization logic in a single place?